File: split-mix-into-songs.py
import os
import re
import ffmpeg
import shutil

def read_description_file(filepath):
    with open(filepath, 'r') as f:
        return f.readlines()

def extract_timestamps(lines):
    pattern = re.compile(r"(\d\d?:\d\d?:?\d?\d?)\s+(.+?)\s+-\s+(.+)")
    timestamps = []
    for line in lines:
        line = line.strip().lower()
        if not line:
            continue
        match = pattern.match(line)
        if match:
            timestamps.append((match.group(1), match.group(2), match.group(3)))
    return timestamps

def get_audio_duration(filepath):
    probe = ffmpeg.probe(filepath)
    return float(probe['streams'][0]['duration'])

def time_to_seconds(time_str):
    parts = list(map(int, time_str.split(':')))
    if len(parts) == 2:
        m, s = parts
        h = 0
    else:
        h, m, s = parts
    return round(h * 3600 + m * 60 + s)

def split_audio(timestamps, input_audio, output_dir, done_dir, dry_run=False):
    duration = get_audio_duration(input_audio)
    os.makedirs(output_dir, exist_ok=True)

    for i, (start, artist, title) in enumerate(timestamps):
        artist = artist.replace('/', '').replace('\\', '')
        title = title.replace('/', '').replace('\\', '')
        start_sec = time_to_seconds(start)
        end_sec = time_to_seconds(timestamps[i + 1][0]) if i + 1 < len(timestamps) else duration

        output_file = os.path.join(output_dir, f"{artist} - {title} ({i}).mp3")
        if (os.path.exists(output_file)):
            print(f"!! [E] File already exists: {output_file}")
        else: 
            command = (
                ffmpeg
                .input(input_audio, ss=start_sec, t=end_sec - start_sec)
                .output(output_file, codec='copy')
            )
            
            if dry_run:
                print("-- [I] Dry run:", command.compile())
            else:
                try:
                    command.run()
                except Exception as e:
                    print(f"!! [E] ffmpeg: {e}")

def process_description_files(input_dir, output_dir, done_dir, dry_run=False):
    description_files = [f for f in os.listdir(input_dir) if f.endswith(".description")]
    description_files.sort()
    total_files = len(description_files)

    for index, filename in enumerate(description_files):
        try:
            print(f"\n\n-- [I] Processing file {index + 1}/{total_files}: {filename}")

            filepath = os.path.join(input_dir, filename)
            lines = read_description_file(filepath)
            timestamps = extract_timestamps(lines)
            num_songs = len(timestamps)
            print(f"-- [I] Found {num_songs} songs.")
            if num_songs > 0:
                audio_filename = os.path.splitext(filename)[0] + ".mp3"
                audio_filepath = os.path.join(input_dir, audio_filename)

                split_audio(timestamps, audio_filepath, output_dir, done_dir, dry_run=dry_run)

                desc_out = os.path.join(done_dir, filename)
                audio_out = os.path.join(done_dir, audio_filename)
                print(f"-- [I] Moving file to {desc_out}")
                print(f"-- [I] Moving file to {audio_out}")
                if dry_run:
                    print("-- [I] Dry run:", f"shutil.move({filepath}, {desc_out})")
                    print("-- [I] Dry run:", f"shutil.move({audio_filepath}, {audio_out})")
                else:
                    shutil.move(filepath, desc_out)
                    shutil.move(audio_filepath, audio_out)
                print(f"++ [D] Finished processing file {index + 1}/{total_files}: {filename}\n")
            else:
                print(f"!! [E] No songs in description; skipping.")
        except Exception as e:
            print(f"!! [E] Error processing file {filename}: {e}")


if __name__ == "__main__":
    base_dir = os.getcwd()
    input_dir = os.path.join(base_dir, "mixes")
    output_dir = os.path.join(base_dir, "songs")
    done_dir = os.path.join(base_dir, "processed")
    os.makedirs(output_dir, exist_ok=True)
    os.makedirs(done_dir, exist_ok=True)
    process_description_files(input_dir, output_dir, done_dir, dry_run=False)

File: src/split-mix-into-songs.ts
import * as fs from 'fs';
import * as path from 'path';
import ffmpeg from 'fluent-ffmpeg';

interface Timestamp {
  start: string;
  artist: string;
  title: string;
}

function readDescriptionFile(filepath: string): string[] {
  return fs.readFileSync(filepath, 'utf8').split('\n');
}

function extractTimestamps(lines: string[]): Timestamp[] {
  const rangePattern = /(\d\d?:\d\d?:?\d?\d?)\s*-\s*\d\d?:\d\d?:?\d?\d?\s+(.+?)\s+-\s+(.+)/;
  const timestampPattern = /(\d\d?:\d\d?:?\d?\d?)\s+(.+)/;
  const timestamps: Timestamp[] = [];
  
  for (const line of lines) {
    const trimmedLine = line.trim();
    if (!trimmedLine) continue;
    
    let match = rangePattern.exec(trimmedLine);
    if (match) {
      timestamps.push({
        start: match[1],
        artist: match[2],
        title: match[3]
      });
    } else {
      match = timestampPattern.exec(trimmedLine);
      if (match) {
        const titlePart = match[2];
        const dashIndex = titlePart.lastIndexOf(' - ');
        
        if (dashIndex > 0) {
          timestamps.push({
            start: match[1],
            artist: titlePart.substring(0, dashIndex).trim(),
            title: titlePart.substring(dashIndex + 3).trim()
          });
        } else {
          timestamps.push({
            start: match[1],
            artist: '',
            title: titlePart.trim()
          });
        }
      }
    }
  }
  
  return timestamps;
}

function getAudioDuration(filepath: string): Promise<number> {
  return new Promise((resolve, reject) => {
    ffmpeg.ffprobe(filepath, (err, metadata) => {
      if (err) {
        reject(err);
        return;
      }
      resolve(metadata.format.duration || 0);
    });
  });
}

function timeToSeconds(timeStr: string): number {
  const parts = timeStr.split(':').map(Number);
  let h = 0, m = 0, s = 0;
  
  if (parts.length === 2) {
    [m, s] = parts;
  } else {
    [h, m, s] = parts;
  }
  
  return Math.round(h * 3600 + m * 60 + s);
}

async function splitAudio(
  timestamps: Timestamp[],
  inputAudio: string,
  outputDir: string,
  doneDir: string,
  dryRun: boolean = false
): Promise<void> {
  const duration = await getAudioDuration(inputAudio);
  
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  for (let i = 0; i < timestamps.length; i++) {
    const { start, artist, title } = timestamps[i];
    const cleanArtist = artist.replace(/[/\\]/g, '');
    const cleanTitle = title.replace(/[/\\]/g, '');
    const startSec = timeToSeconds(start);
    const endSec = i + 1 < timestamps.length 
      ? timeToSeconds(timestamps[i + 1].start) 
      : duration;

    const outputFile = path.join(outputDir, `${cleanArtist} - ${cleanTitle} (${i}).mp3`);
    
    if (fs.existsSync(outputFile)) {
      console.log(`!! [E] File already exists: ${outputFile}`);
    } else {
      if (dryRun) {
        console.log(`-- [I] Dry run: ffmpeg -i "${inputAudio}" -ss ${startSec} -t ${endSec - startSec} -c copy "${outputFile}"`);
      } else {
        try {
          await new Promise<void>((resolve, reject) => {
            ffmpeg(inputAudio)
              .seekInput(startSec)
              .duration(endSec - startSec)
              .audioCodec('copy')
              .output(outputFile)
              .on('end', () => resolve())
              .on('error', (err: any) => reject(err))
              .run();
          });
        } catch (error) {
          console.log(`!! [E] ffmpeg: ${error}`);
        }
      }
    }
  }
}

async function processDescriptionFiles(
  inputDir: string,
  outputDir: string,
  doneDir: string,
  dryRun: boolean = false
): Promise<void> {
  const descriptionFiles = fs.readdirSync(inputDir)
    .filter(f => f.endsWith('.description'))
    .sort();
  
  const totalFiles = descriptionFiles.length;

  for (let index = 0; index < descriptionFiles.length; index++) {
    const filename = descriptionFiles[index];
    
    try {
      console.log(`\n\n-- [I] Processing file ${index + 1}/${totalFiles}: ${filename}`);

      const filepath = path.join(inputDir, filename);
      const lines = readDescriptionFile(filepath);
      const timestamps = extractTimestamps(lines);
      const numSongs = timestamps.length;
      
      console.log(`-- [I] Found ${numSongs} songs.`);
      
      if (numSongs > 0) {
        const audioFilename = path.basename(filename, '.description') + '.mp3';
        const audioFilepath = path.join(inputDir, audioFilename);

        await splitAudio(timestamps, audioFilepath, outputDir, doneDir, dryRun);

        const descOut = path.join(doneDir, filename);
        const audioOut = path.join(doneDir, audioFilename);
        
        console.log(`-- [I] Moving file to ${descOut}`);
        console.log(`-- [I] Moving file to ${audioOut}`);
        
        if (dryRun) {
          console.log(`-- [I] Dry run: fs.rename(${filepath}, ${descOut})`);
          console.log(`-- [I] Dry run: fs.rename(${audioFilepath}, ${audioOut})`);
        } else {
          fs.renameSync(filepath, descOut);
          fs.renameSync(audioFilepath, audioOut);
        }
        
        console.log(`++ [D] Finished processing file ${index + 1}/${totalFiles}: ${filename}\n`);
      } else {
        console.log(`!! [E] No songs in description; skipping.`);
      }
    } catch (error) {
      console.log(`!! [E] Error processing file ${filename}: ${error}`);
    }
  }
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const dryRun = args.includes('--dry-run') || args.includes('-d');
  
  const baseDir = process.cwd();
  const inputDir = path.join(baseDir, 'mixes');
  const outputDir = path.join(baseDir, 'songs');
  const doneDir = path.join(baseDir, 'processed');
  
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }
  
  if (!fs.existsSync(doneDir)) {
    fs.mkdirSync(doneDir, { recursive: true });
  }
  
  console.log(`Running in ${dryRun ? 'dry-run' : 'live'} mode`);
  await processDescriptionFiles(inputDir, outputDir, doneDir, dryRun);
}

if (require.main === module) {
  main().catch(console.error);
}
